File: packages/scheduling-engine/src/constraints/operator-pool.ts
// ═══════════════════════════════════════════════════════════
//  INCOMPOL PLAN — OperatorPool Constraint (Labor-Group-Based)
//  Per-window, per-labor-group operator capacity check.
//
//  Uses WorkforceConfig with LaborWindow[] per labor group
//  (Grandes, Medias). Capacity varies by time window:
//    07:00-15:30, 15:30-16:00, 16:00-00:00
//
//  Mode: ADVISORY — operator capacity warns but never blocks.
//
//  Peak concurrent model: each machine's peak operator demand
//  in a window is tracked; group total = sum of machine peaks.
//
//  R8: Machines not mapped → unmapped=true (flagged, not blocked).
//  R9: Blocks spanning windows → evaluated per segment, WORST wins.
// ═══════════════════════════════════════════════════════════

import type { LaborWindow, WorkforceConfig } from '../types/workforce.js';

// ── Result type ──

/** Result of checking operator capacity */
export interface OperatorCheckResult {
  /** Whether there is enough capacity (across all overlapping windows) */
  hasCapacity: boolean;
  /** Available operator slots (minimum across windows) */
  available: number;
  /** Resolved labor group (undefined if machine not mapped) */
  laborGroup: string | undefined;
  /** Whether the machine is not mapped to any labor group (R8) */
  unmapped: boolean;
  /** Worst window shortage: max(0, peakNeed - capacity) across windows */
  worstWindowShortage: number;
}

// ── Factory ──

/**
 * Creates an OperatorPool constraint instance using labor-group-based configuration.
 *
 * Machines are mapped to labor groups via config.machineToLaborGroup.
 * Capacity is looked up per labor group per time window.
 * Machines not in the mapping are unconstrained but flagged as unmapped (R8).
 *
 * Uses a "peak concurrent" model: each machine's peak operator demand
 * in a window is tracked, and the group total is the sum of peaks.
 *
 * @param config - WorkforceConfig with laborGroups and machine-to-laborGroup mapping
 */
export function createOperatorPool(config: WorkforceConfig) {
  /**
   * Resolve labor group for a machine. Returns undefined if not mapped.
   */
  function resolveLaborGroup(machineId: string): string | undefined {
    return config.machineToLaborGroup[machineId];
  }

  /**
   * Get all windows that overlap with a time range [startMin, endMin).
   */
  function getOverlappingWindows(
    laborGroup: string,
    startMin: number,
    endMin: number,
  ): LaborWindow[] {
    const windows = config.laborGroups[laborGroup];
    if (!windows) return [];
    return windows.filter((w) => startMin < w.end && endMin > w.start);
  }

  // ── State ──

  /** Peak operators per machine per window: key = "di:windowStart:machineId" */
  const machPeak: Record<string, number> = {};

  /** Sum of peaks per group per window: key = "di:windowStart:laborGroup" */
  const groupTotal: Record<string, number> = {};

  return {
    /**
     * Check if there is enough operator capacity for a block [startMin, endMin).
     * Does NOT modify state — use book() to commit.
     *
     * R9: If the block spans multiple windows, checks each segment separately
     * and returns the WORST result.
     *
     * R8: Machines not mapped return unmapped=true.
     *
     * @param di        - Day index (0-based)
     * @param startMin  - Block start minute within day
     * @param endMin    - Block end minute within day
     * @param operators - Number of operators needed
     * @param machineId - Machine ID (resolved to labor group internally)
     * @returns OperatorCheckResult
     */
    checkCapacity(
      di: number,
      startMin: number,
      endMin: number,
      operators: number,
      machineId: string,
    ): OperatorCheckResult {
      const laborGroup = resolveLaborGroup(machineId);
      if (!laborGroup) {
        return {
          hasCapacity: true,
          available: Infinity,
          laborGroup: undefined,
          unmapped: true,
          worstWindowShortage: 0,
        };
      }

      const windows = getOverlappingWindows(laborGroup, startMin, endMin);
      if (windows.length === 0) {
        return {
          hasCapacity: true,
          available: Infinity,
          laborGroup,
          unmapped: false,
          worstWindowShortage: 0,
        };
      }

      let overallHasCapacity = true;
      let minAvailable = Infinity;
      let worstShortage = 0;

      for (const w of windows) {
        const machKey = `${di}:${w.start}:${machineId}`;
        const groupKey = `${di}:${w.start}:${laborGroup}`;

        const currentMachPeak = machPeak[machKey] || 0;
        const delta = Math.max(0, operators - currentMachPeak);
        const currentGroupTotal = groupTotal[groupKey] || 0;
        const newTotal = currentGroupTotal + delta;

        const available = w.capacity - currentGroupTotal;
        if (available < minAvailable) minAvailable = available;

        if (newTotal > w.capacity) {
          overallHasCapacity = false;
          const shortage = newTotal - w.capacity;
          if (shortage > worstShortage) worstShortage = shortage;
        }
      }

      return {
        hasCapacity: overallHasCapacity,
        available: Math.max(0, minAvailable),
        laborGroup,
        unmapped: false,
        worstWindowShortage: worstShortage,
      };
    },

    /**
     * Simplified capacity check (returns boolean).
     */
    hasCapacity(
      di: number,
      startMin: number,
      endMin: number,
      operators: number,
      machineId: string,
    ): boolean {
      return this.checkCapacity(di, startMin, endMin, operators, machineId).hasCapacity;
    },

    /**
     * Book operator demand. Updates the peak model state.
     * Call this AFTER deciding to schedule the block.
     *
     * R9: Books into each overlapping window segment separately.
     *
     * @param di        - Day index (0-based)
     * @param startMin  - Block start minute within day
     * @param endMin    - Block end minute within day
     * @param operators - Number of operators
     * @param machineId - Machine ID (resolved to labor group internally)
     */
    book(di: number, startMin: number, endMin: number, operators: number, machineId: string): void {
      const laborGroup = resolveLaborGroup(machineId);
      if (!laborGroup) return;

      const windows = getOverlappingWindows(laborGroup, startMin, endMin);

      for (const w of windows) {
        const machKey = `${di}:${w.start}:${machineId}`;
        const currentMachPeak = machPeak[machKey] || 0;
        const delta = Math.max(0, operators - currentMachPeak);
        machPeak[machKey] = Math.max(currentMachPeak, operators);

        const groupKey = `${di}:${w.start}:${laborGroup}`;
        groupTotal[groupKey] = (groupTotal[groupKey] || 0) + delta;
      }
    },

    /**
     * Get current usage for a labor group/window/day.
     *
     * @param di          - Day index
     * @param windowStart - Window start minute
     * @param laborGroup  - Labor group ID
     * @returns Current total operator demand (sum of machine peaks)
     */
    getCurrentUsage(di: number, windowStart: number, laborGroup: string): number {
      return groupTotal[`${di}:${windowStart}:${laborGroup}`] || 0;
    },

    /**
     * Resolve the labor group for a machine.
     */
    getLaborGroup(machineId: string): string | undefined {
      return resolveLaborGroup(machineId);
    },

    /** Clear all state (for test or reset) */
    clear(): void {
      for (const key of Object.keys(machPeak)) delete machPeak[key];
      for (const key of Object.keys(groupTotal)) delete groupTotal[key];
    },
  };
}

/** The return type of createOperatorPool() */
export type OperatorPool = ReturnType<typeof createOperatorPool>;

File: apps/backend/src/domain/scheduling/constraints/operator_pool.py
"""OperatorPool constraint — capacidade operadores por turno/área.

Port of constraints/operator-pool.ts.
Mode: ADVISORY — warns but never blocks.
"""

from __future__ import annotations

from dataclasses import dataclass

from ..types import WorkforceConfig


@dataclass
class OperatorCheckResult:
    has_capacity: bool
    current_usage: int
    capacity: int
    shortage: int
    labor_group: str
    is_warning: bool


DEFAULT_LABOR_GROUPS: dict[str, list[dict]] = {
    "Grandes": [
        {"start": 420, "end": 930, "capacity": 6},
        {"start": 930, "end": 960, "capacity": 6},
        {"start": 960, "end": 1440, "capacity": 5},
    ],
    "Medias": [
        {"start": 420, "end": 930, "capacity": 9},
        {"start": 930, "end": 960, "capacity": 8},
        {"start": 960, "end": 1440, "capacity": 4},
    ],
}

DEFAULT_MACHINE_TO_GROUP: dict[str, str] = {
    "PRM019": "Grandes",
    "PRM031": "Grandes",
    "PRM039": "Grandes",
    "PRM043": "Grandes",
    "PRM042": "Medias",
}


class OperatorPool:
    """Labor group operator capacity — advisory, never blocks."""

    def __init__(self, config: WorkforceConfig | None = None) -> None:
        if config and config.labor_groups:
            self._labor_groups = {
                g: [{"start": w.start, "end": w.end, "capacity": w.capacity} for w in windows]
                for g, windows in config.labor_groups.items()
            }
            self._machine_map = dict(config.machine_to_labor_group)
        else:
            self._labor_groups = DEFAULT_LABOR_GROUPS
            self._machine_map = dict(DEFAULT_MACHINE_TO_GROUP)

        # Usage tracking: (day, window_start, labor_group) → {machine → peak_operators}
        self._usage: dict[tuple[int, int, str], dict[str, int]] = {}

    def get_labor_group(self, machine_id: str) -> str:
        return self._machine_map.get(machine_id, "")

    def check_capacity(
        self,
        day_idx: int,
        start_min: int,
        end_min: int,
        operators: int,
        machine_id: str,
    ) -> OperatorCheckResult:
        """Check if adding operators would exceed capacity."""
        group = self.get_labor_group(machine_id)
        if not group:
            return OperatorCheckResult(
                has_capacity=True,
                current_usage=0,
                capacity=99,
                shortage=0,
                labor_group="",
                is_warning=False,
            )

        windows = self._labor_groups.get(group, [])
        worst_shortage = 0
        worst_cap = 99
        worst_usage = 0

        for w in windows:
            if start_min >= w["end"] or end_min <= w["start"]:
                continue
            cap = w["capacity"]
            key = (day_idx, w["start"], group)
            machine_usage = self._usage.get(key, {})
            total = sum(machine_usage.values())
            new_total = total + operators
            shortage = max(new_total - cap, 0)
            if shortage > worst_shortage:
                worst_shortage = shortage
                worst_cap = cap
                worst_usage = total

        return OperatorCheckResult(
            has_capacity=worst_shortage == 0,
            current_usage=worst_usage,
            capacity=worst_cap,
            shortage=worst_shortage,
            labor_group=group,
            is_warning=worst_shortage > 0,
        )

    def book(
        self,
        day_idx: int,
        start_min: int,
        end_min: int,
        operators: int,
        machine_id: str,
    ) -> None:
        """Record operator usage."""
        group = self.get_labor_group(machine_id)
        if not group:
            return
        windows = self._labor_groups.get(group, [])
        for w in windows:
            if start_min >= w["end"] or end_min <= w["start"]:
                continue
            key = (day_idx, w["start"], group)
            if key not in self._usage:
                self._usage[key] = {}
            current = self._usage[key].get(machine_id, 0)
            self._usage[key][machine_id] = max(current, operators)

    def get_current_usage(self, day_idx: int, window_start: int, labor_group: str) -> int:
        key = (day_idx, window_start, labor_group)
        machine_usage = self._usage.get(key, {})
        return sum(machine_usage.values())

    def clear(self) -> None:
        self._usage.clear()
